### PR TITLE
ADM-444[Report] [backend] Change failure rate calculation

### PR DESCRIPTION
**Pull request by fengxinhou**
## Summary

Change failure rate calculation

## After
Postman:
Screen Shot 2023-05-12 at 12 20 30 PM

frontend:
Screen Shot 2023-05-12 at 11 34 54 AM

### CODE REVIEW

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java [hunk 1]
```diff
@@ -5,10 +5,7 @@
 import heartbeat.controller.report.dto.request.BuildKiteSetting;
 import heartbeat.controller.report.dto.request.GenerateReportRequest;
 import heartbeat.controller.report.dto.request.JiraBoardSetting;
-import heartbeat.controller.report.dto.response.AvgDeploymentFrequency;
-import heartbeat.controller.report.dto.response.DeploymentFrequency;
-import heartbeat.controller.report.dto.response.ReportResponse;
-import heartbeat.controller.report.dto.response.Velocity;
+import heartbeat.controller.report.dto.response.*;
```
Comment (wangwang-zhang) [inline backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:8]: remove *

File: backend/src/main/java/heartbeat/client/dto/pipeline/buildkite/BuildKiteBuildInfo.java [hunk 2]
```diff
@@ -26,9 +27,15 @@ public class BuildKiteBuildInfo {
 
 	private int number;
 
-	public DeployInfo mapToDeployInfo(String step, String states) {
+	public DeployInfo mapToDeployInfo(String step, String states, String startTime, String endTime) {
+		Instant startDate = Instant.ofEpochMilli(Long.parseLong(startTime));
+		Instant endDate = Instant.ofEpochMilli(Long.parseLong(endTime));
 		BuildKiteJob job = this.jobs.stream()
 			.filter(item -> Objects.equals(item.getName(), step) && Objects.equals(states, item.getState()))
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/client/dto/pipeline/buildkite/BuildKiteBuildInfo.java:34]: states to state

File: backend/src/main/java/heartbeat/service/report/ChangeFailureRateCalculator.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.stereotype.Component;
+
+import java.text.DecimalFormat;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Component
+@Log4j2
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/service/report/ChangeFailureRateCalculator.java:16]: why add @Log4j2, but you don't use it, check @RequiredArgsConstructor also

File: backend/src/main/java/heartbeat/service/report/ChangeFailureRateCalculator.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+public class ChangeFailureRateCalculator {
+
+	public static final String FORMAT_4_DECIMALS = "0.0000";
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/service/report/ChangeFailureRateCalculator.java:19]: change public to private

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 5]
```diff
@@ -67,11 +69,14 @@ public ReportResponse generateReporter(GenerateReportRequest request) {
...
 					reportResponse.setDeploymentFrequency(deploymentFrequency.calculate(deployTimesList,
 							Long.parseLong(request.getStartTime()), Long.parseLong(request.getEndTime())));
+				case "change failure rate" ->
+					reportResponse.setChangeFailureRate(changeFailureRate.calculate(deployTimesList));
 				default -> {
 					// TODO
 				}
 			}
 		});
+		log.info("*********reportResponse:{}", reportResponse);
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:79]: Is this log ok?

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java [hunk 6]
```diff
@@ -66,12 +67,20 @@ class BuildKiteServiceTest {
 			<https://api.buildkite.com/v2/organizations/test_org_id/pipelines/test_pipeline_id/builds?per_page=100&page=2>; rel="next"
 			""";
 
+	public static final String PASSED_STATE = "passed";
+
+	public static final String FAILED_STATE = "failed";
```
Comment (wangwang-zhang) [inline backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java:72]: private ,please

File: backend/src/test/java/heartbeat/service/report/CalculateChangeFailureRateTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package heartbeat.service.report;
+
+import heartbeat.client.dto.pipeline.buildkite.DeployInfo;
+import heartbeat.client.dto.pipeline.buildkite.DeployTimes;
+import heartbeat.controller.report.dto.response.*;
```
Comment (wangwang-zhang) [inline backend/src/test/java/heartbeat/service/report/CalculateChangeFailureRateTest.java:5]: remove *

File: backend/src/test/java/heartbeat/service/report/CalculateChangeFailureRateTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+@MockitoSettings(strictness = Strictness.LENIENT)
+class CalculateChangeFailureRateTest {
+
+	public static final String JOB_FINISH_TIME_2022 = "2022-09-08T22:45:33.981Z";
+
+	public static final String JOB_FINISH_TIME_2023 = "2023-09-08T22:45:33.981Z";
+
+	public static final String PASSED_STATE = "passed";
+
+	public static final String FAILED_STATE = "failed";
```
Comment (wangwang-zhang) [inline backend/src/test/java/heartbeat/service/report/CalculateChangeFailureRateTest.java:28]: private ?

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java (hunk 1 above)
Comment (fengxinhou) [reply backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:8]: done

**Review by wangwang-zhang — APPROVED** [review]
LGTM

**Review by TingLiMio — APPROVED** [review]
LGFM

**Pull request merged**